### ISSUE TITLE
Error running the Java sample with intel-mkl 2020.2.254-1

### ISSUE THREAD

**Issue by jeremi** [lagadic/visp#806]
On arch linux, I'm getting an issue running the java tutorial for `ApriltagDetection.java`.
- I'm using the latest `master` ( f42f94d ).
- I have intel-mkl `2020.2.254-1`. My processor is an AMD, but I don't think this is the issue.
- my [ViSP-third-party.txt](https://gist.github.com/jeremi/895cf384c3e09cb31af551813d007f88)

```
TAG_CIRCLE49h12, TAG_CUSTOM48h12, TAG_STANDARD41h12 and TAG_STANDARD52h13 are disabled.
INTEL MKL ERROR: /opt/intel/mkl/lib/intel64/libmkl_def.so: undefined symbol: mkl_sparse_optimize_bsr_trsm_i8.
Intel MKL FATAL ERROR: Cannot load libmkl_def.so.

Process finished with exit code 2
```

**Comment by fspindle** [comment]
Same issue occurs on Ubuntu 18.04 with `l_mkl_2020.0.166.tgz`

1. Note: if Eclipse is not launched from the terminal, running the example from "Run > Run" menu lead to: 

    ````
    Exception in thread "main" java.lang.UnsatisfiedLinkError: /home/fspindle/visp_ws/visp-build/lib/libvisp_java331.so: libmkl_intel_lp64.so: cannot open shared object file: No such file or directory
    ````

2. Starting Eclipse from terminal like `$./eclipse` allows to run the application that opens the interface. But then, after selection of a 36h11 image, pressing `Detect` button lead to the error message:
    ```
    INTEL MKL ERROR: /opt/intel/compilers_and_libraries_2020.0.166/linux/mkl/lib/intel64_lin/libmkl_avx2.so: undefined symbol: mkl_sparse_optimize_bsr_trsm_i8.
    Intel MKL FATAL ERROR: Cannot load libmkl_avx2.so or libmkl_def.so.
    ```

**Comment by fspindle** [comment]
A quick trial shows that this issue is fixed when `cmake/FindMKL.cmake` file is modified to find `mkl_rt` instead of `mkl_core`+`mkl_intel_thread`+`mkl_sequential`

**Comment by s-trinh** [comment]
> I have intel-mkl `2020.2.254-1`. My processor is an AMD, but I don't think this is the issue.

Intel has been (is still?) notorious for anti-competitive practices (e.g. [against AMD](https://en.wikipedia.org/wiki/Advanced_Micro_Devices,_Inc._v._Intel_Corp.)).

Specifically: [Will Intel be forced to remove the "cripple AMD" function from their compiler?](https://www.agner.org/optimize/blog/read.php?i=49) where ICC (the Intel C++ compiler) did not generate the same level of optimizations regardless the CPU vendor while the ISA (instruction set architecture) is the same.
See also: [Overview of CPU dispatching in Intel software](https://www.agner.org/optimize/blog/read.php?i=49#121) for an overview in 2010.

Don't know what is the current performance of MKL on AMD nowadays but a quick search gives: [OpenBLAS is faster than Intel MKL on AMD Hardware (Ryzen)](https://discourse.julialang.org/t/openblas-is-faster-than-intel-mkl-on-amd-hardware-ryzen/8033), so...

Found also this: [Boosting Intel MKL on AMD Ryzen processors](https://simon-martin.net/2020/03/01/boosting-intel-mkl-on-amd-ryzen-processors/):
> Hi Ralf, thanks for the remark. I can confirm this – in MKL 2020 update 1, Intel pulled the plug for the debug mode. This is very bad with regard to upcoming Matlab releases which will ship with MKL 2020.0.1 or higher. AMD users of NumPy and TensorFlow should imo rather rely on OpenBLAS anyway.

Where it is not possible anymore with `Intel MKL 2020 update 1` to workaround the CPU vendor ID check...

I would build from source OpenBLAS on an AMD platform.

---

> Intel MKL FATAL ERROR: Cannot load libmkl_avx2.so or libmkl_def.so.

I remember solving the issue by doing this in a terminal before launching Eclipse from the terminal: `export LD_PRELOAD=/opt/intel/mkl/lib/intel64/libmkl_core.so:/opt/intel/mkl/lib/intel64/libmkl_sequential.so` (path to be checked).

I found also this: [Intel® Math Kernel Library (Intel MKL) Linkage and Distribution - Quick Reference Guide](https://software.intel.com/content/www/us/en/develop/articles/intel-math-kernel-library-intel-mkl-linkage-and-distribution-quick-reference-guide.html)

**Comment by jeremi** [comment]
I'm not sure it is possible with CMake, but we could disable the selection of mkl on AMD and search for `mkl_rt`? That way, if someone is on an AMD platform, it would take openBLAS if available. Right now as MKL is the first, it will always select it.

**Comment by fspindle** [comment]
1. I was able to reproduce the issue on Ubuntu 18.04 with an Intel processor. I don't think that the issue is related to the processor type
2. Introducing a trick to turn MKL usage off on AMD will not resolve the issue on Intel. Anyway, for me it is the responsability of the user to know which 3rd party is optimal for a given platform
3. I also tried what you mention, but it doesn't work for me
    > I remember solving the issue by doing this in a terminal before launching Eclipse from the terminal: `export LD_PRELOAD=/opt/intel/mkl/lib/intel64/libmkl_core.so:/opt/intel/mkl/lib/intel64/libmkl_sequential.so` (path to be checked).
4. The solution I see is to use the Single Dynamic Library ie `mkl_rt.so`

**Comment by jeremi** [comment]
yes, though for 2. I have MKL installed for another lib and that require to modify the CMake file to skip it. I have not seen a parameter when running `ccmake` to disable the use of MKL.

maybe a `USE_MKL` switch be useful for people who do not know how to modify the cmake config? But then it might be good to have the same switch for the alternatives.

**Comment by fspindle** [comment]
Yes, I tought that it could be also a good idea to add the possibility to disable `USE_BLAS/LAPACK` setting this var to `OFF`.  In that case  `WITH_LAPACK` should be set automatically to `ON`.

**Comment by fspindle** [comment]
There are some regression during execution on Ubuntu when using `mkl_rt` instead of `mkl_core` + `mkl_intel_thread` + `mkl_sequential`

To keep a feedback:
```
$ ldd mbtEdgeTracking | grep mkl
libmkl_rt.so => /opt/intel/compilers_and_libraries_2020.0.166/linux/mkl/lib/intel64_lin/libmkl_rt.so (0x00007fa7ef740000)
$ ./mbtEdgeTracking -c
...
lod : min polygon area threshold : 2500 (default)
Catch an exception: Error [8]:	The algorithm computing SVD failed to converge.
```
while with the current `FindMKL.cmake` file the same example is working:
```
$ ldd mbtEdgeTracking |grep mkl
	libmkl_intel_lp64.so => /opt/intel/compilers_and_libraries_2020.0.166/linux/mkl/lib/intel64_lin/libmkl_intel_lp64.so (0x00007f513399d000)
	libmkl_sequential.so => /opt/intel/compilers_and_libraries_2020.0.166/linux/mkl/lib/intel64_lin/libmkl_sequential.so (0x00007f5132385000)
	libmkl_core.so => /opt/intel/compilers_and_libraries_2020.0.166/linux/mkl/lib/intel64_lin/libmkl_core.so (0x00007f512e065000)
$ ./mbtEdgeTracking -c
-> ok
```

**Comment by fspindle** [comment]
@jeremi Could you please test PR #811 turning `USE_BLAS/LAPACK` option to OFF.
Using a command line similar to the following should disable MKL usage in your case:
```
$ cmake ../visp -DUSE_BLAS/LAPACK=OFF
```

**Comment by jeremi** [comment]
When I run the command as you gave, it builds, but in the file [ViSP-third-party.txt](https://gist.github.com/jeremi/27d6de6ca831b1be429a1cf14420a116) I can still see:
```
Mathematics: 
    Blas/Lapack:                 yes
    \- Use MKL:                  no
    \- Use OpenBLAS:             no
    \- Use Atlas:                no
    \- Use Netlib:               no
    \- Use GSL:                  no
    \- Use Lapack (built-in):    yes (ver 3.2.1)
    Use Eigen3:                  yes (ver 3.3.7)
    Use OpenCV:                  yes (ver 4.4.0)
```

I did run the tests as well and they are all passing.

**Comment by fspindle** [comment]
The `ViSP-third-party.txt`  content is as expected. MKL, OpenBLAS, Atlas, Netlib and GSL are turned OFF using `$ cmake ../visp -DUSE_BLAS/LAPACK=OFF` while  Lapack (built-in) is turned ON.

**Comment by fspindle** [comment]
Fixed in #811

**Comment by s-trinh** [comment]
@fspindle 

FYI.
I have tested the `LD_PRELOAD` method and I am able to run the MBT Java sample with latest MKL on Ubuntu 16.04 and 20.04.
I have also `source /opt/intel/compilers_and_libraries_2020/linux/bin/compilervars.sh -arch intel64 -platform linux` in the `.bashrc`
